### DISCUSSION TITLE
Redirect stdout/stderr with third-party console tools

### DISCUSSION THREAD

**Discussion by lonix1** [spectreconsole/spectre.console#965] [Q&A]
I'm using a third-party console tool ([System.CommandLine](https://github.com/dotnet/command-line-api)), with SpectreConsole for logging (obviously... there's nothing quite like it).

I want to write error messages to `stdout`. I can do that with SC.L, but not when I combine both libraries. The closest I found was [this issue](https://github.com/spectreconsole/spectre.console/issues/359) but I'm unsure how to use the info there practically.

I'm assuming there are other people who use these two together - do you have any working code or tips to share?

**Comment by lonix1** [comment]
Just found something that works...

I created a new logger which logs to `stdout`, and SC.L supplies the `TextWriter`:
````cs
// get text write from SC.L
using var writer = context.Console.Error.CreateTextWriter();

// create new Spectre logger, just for stderr
var errConsole = AnsiConsole.Create(new AnsiConsoleSettings {
  Ansi                   = console.Profile.Capabilities.Ansi ? AnsiSupport.Yes : AnsiSupport.Detect,
  ColorSystem            = (ColorSystemSupport)console.Profile.Capabilities.ColorSystem,
  //Enrichment           = ??? (***)
  //EnvironmentVariables = ??? (***)
  ExclusivityMode        = console.ExclusivityMode,
  Interactive            = console.Profile.Capabilities.Interactive ? InteractionSupport.Yes : InteractionSupport.Detect,
  Out                    = new AnsiConsoleOutput(writer),
  // ...what else to do here? (***)
});

console.MarkupLine("[green]stdout[/]");
errConsole.MarkupLine("[red]stderr[/]");
````

Test:
````sh
$ myapp                    # prints both in colour
$ myapp 1>/dev/null        # prints red 'stderr'
$ myapp 2>/dev/null        # prints green 'stdout'
````

..."Works for me", on my terminal.

- Have I done this correctly? (I was unsure whether to set `console.Profile.Out` instead.)
- It complains with *Spectre1010: Favor the use of the instance of AnsiConsole over the static helper* - can I ignore that or am I using the API incorrectly?
- I populated the new console with settings `(***)` from the normal `stdout` console - are there other settings I should map?

**Comment by phil-scott-78** [comment]
Hmmm, not 100% sure what the SC.L is referring too. If it is a logging nuget package out there, we don't own that so this might be better suited asking that package owner.

**Reply by lonix1** [reply]
It's [System.CommandLine](https://github.com/dotnet/command-line-api). Once released it will be the standard (it's from Microsoft). And this library is the standard for console printing. So many people will use them together.
